core/drivers/crypto/caam/caam_rng.c

```diff
@@ -0,0 +1,575 @@
+	}
+
+	return CAAM_NO_ERROR;
+}
+
+/* Free module private data */
+static void do_free(void)
+{
+	struct rngdata *rng = NULL;
+	unsigned int idx = 0;
+
+	if (rng_privdata) {
+		for (idx = 0; idx < RNG_DATABUF_NB; idx++) {
+			rng = &rng_privdata->databuf[idx];
+
+			/* Check if there is a Job ongoing to cancel it */
+			if (atomic_load_u32(&rng->status) == DATA_ONGOING)
+				caam_jr_cancel(rng->job_id);
+
+			caam_free_desc(&rng->jobctx.desc);
+			caam_free(rng->data);
+			rng->data = NULL;
+		}
+
+		caam_free(rng_privdata);
+		rng_privdata = NULL;
+	}
+}
+
+#ifdef CFG_CRYPTO_RNG_HW
+/*
+ * RNG data generation job ring callback completion
+ *
+ * @jobctx      RNG data JR Job Context
+ */
+static void rng_data_done(struct caam_jobctx *jobctx)
+{
+	struct rngdata *rng = jobctx->context;
+
+	RNG_TRACE("RNG Data id 0x%08" PRIx32 " done with status 0x%" PRIx32,
+		  rng->job_id, jobctx->status);
+
+	if (JRSTA_SRC_GET(jobctx->status) == JRSTA_SRC(NONE)) {
+		atomic_store_u32(&rng->status, DATA_OK);
+
+		/* Invalidate the data buffer to ensure software got it */
+		cache_operation(TEE_CACHEINVALIDATE, rng->data, rng->size);
+	} else {
+		RNG_TRACE("RNG Data completion in error 0x%" PRIx32,
+			  jobctx->status);
+		atomic_store_u32(&rng->status, DATA_FAILURE);
+	}
+
+	rng->job_id = 0;
+	rng->rdindex = 0;
+}
+
+/*
+ * Prepares the data generation descriptors
+ *
+ * @rng       Reference to the RNG Data object
+ */
+static enum CAAM_Status prepare_gen_desc(struct rngdata *rng)
+{
+	paddr_t paddr = 0;
+	uint32_t *desc = NULL;
+
+	/* Convert the buffer virtual address to physical address */
+	paddr = virt_to_phys(rng->data);
+	if (!paddr)
+		return CAAM_FAILURE;
+
+	desc = rng->jobctx.desc;
+
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc, RNG_GEN_DATA);
+	caam_desc_add_word(desc, FIFO_ST(RNG_TO_MEM, rng->size));
+	caam_desc_add_ptr(desc, paddr);
+
+	RNG_DUMPDESC(desc);
+
+	/* Prepare the job context */
+	rng->jobctx.context = rng;
+	rng->jobctx.callback = rng_data_done;
+	return CAAM_NO_ERROR;
+}
+
+/*
+ * Launches a RNG Data generation
+ *
+ * @rng      RNG Data context
+ */
+static enum CAAM_Status do_rng_start(struct rngdata *rng)
+{
+	enum CAAM_Status ret = CAAM_FAILURE;
+
+	/* Ensure that data buffer is flushed */
+	cache_operation(TEE_CACHEFLUSH, rng->data, rng->size);
+
+	rng->job_id = 0;
+	atomic_store_u32(&rng->status, DATA_EMPTY);
+
+	ret = caam_jr_enqueue(&rng->jobctx, &rng->job_id);
+
+	if (ret == CAAM_PENDING) {
+		atomic_store_u32(&rng->status, DATA_ONGOING);
+		ret = CAAM_NO_ERROR;
+	} else {
+		RNG_TRACE("RNG Job Ring Error 0x%" PRIx32, ret);
+		atomic_store_u32(&rng->status, DATA_FAILURE);
+		ret = CAAM_FAILURE;
+	}
+
+	return ret;
+}
+
+/* Checks if there are random data available */
+static enum CAAM_Status do_check_data(void)
+{
+	enum CAAM_Status ret = CAAM_FAILURE;
+	struct rngdata *rng = NULL;
+	uint32_t wait_jobs = 0;
+	unsigned int idx = 0;
+	unsigned int loop = 4;
```

Comment:
because I estimated that is enough. Returning immediately will be an error because the number of Random Bytes to generate. Waiting more can generate a system crash (linux) because TEE is no responding.

Reply:
ok (resolved)